wts/blueprints/aggregate.py

```diff
@@ -0,0 +1,115 @@
+import asyncio
+import flask
+import httpx
+import time
+from cdiserrors import NotFoundError, UserError
+
+from ..auth import authenticate
+from ..models import db, RefreshToken
+from ..tokens import async_get_access_token
+
+
+blueprint = flask.Blueprint("aggregate", __name__)
+
+
+@blueprint.route("/<path:endpoint>", methods=["GET"])
+async def get_aggregate_response(endpoint):
+    """
+    Proxy GET requests to `endpoint` on each linked commons and return
+    an aggregated response.
+
+    For an authenticated request, each proxied request incudes an access token
+    fetched using the current user's refresh token.
+
+    The size of the aggregated response body can be reduced by supplying a
+    `filters` parameter. If provided, only return JSON key-value pairs whose
+    key is in `filters`. Multiple filters can be specified:
+
+    `GET /aggregate/user/user?filters=authz&filters=username`
+    """
+    # for `GET /aggregate/user/user`, flask sets endpoint to 'user/user'
+    endpoint = f"/{endpoint}"
+    if endpoint not in flask.current_app.config["AGGREGATE_ENDPOINT_ALLOWLIST"]:
+        raise NotFoundError(
+            "supplied endpoint is not configured in the Workspace Token Service aggregate endpoint allowlist"
+        )
+
+    filters = flask.request.args.getlist("filters")
+    parameters = flask.request.args.to_dict()
+    parameters.pop("filters", None)
+
+    if flask.request.headers.get("Authorization"):
+        authenticate(allow_access_token=True)
+        refresh_tokens = (
+            db.session.query(RefreshToken)
+            .filter_by(username=flask.g.user.username)
+            .filter(RefreshToken.expires > int(time.time()))
+            .order_by(RefreshToken.expires.asc())
+        )
+        #  if a user has multiple refresh tokens for the same commons, we want
+        #  the latest one to be used. see
+        #  https://stackoverflow.com/questions/39678672/is-a-python-dict-comprehension-always-last-wins-if-there-are-duplicate-keys
+        refresh_tokens = {
+            flask.current_app.config["OIDC"][rt.idp]["commons_hostname"]: rt
+            for rt in refresh_tokens
+        }
+        access_tokens = await asyncio.gather(
+            *[async_get_access_token(rt) for rt in refresh_tokens.values()]
+        )
+        request_info = [
+            (commons, endpoint, {"Authorization": f"Bearer {access_token}"})
+            if access_token
+            else (commons, None, None)
```

Comment:
My thoughts up to now were that if a user's access info for a commons cannot be determined (i.e. you fail to get an access token), it would be preferable to indicate that access is unknown instead of showing incomplete or inaccurate information. For this reason, I thought a special JSON value should be chosen to indicate that access is unknown (`{}` wasn't the best choice as a special value, since it can't be differentiated from a truly empty Arborist response. If we want to go the route of a special value, maybe `null` would be better).

This is an edge case that we do need to communicate and clarify requirements for.